v0.21.0 fails to compile on raspbian with clang

Getting these errors:-

make[3]: Leaving directory '/home/pi/src/bitcoin-core-v0.21'
/bin/bash ../libtool  --tag=CXX --preserve-dup-deps  --mode=link clang++ -std=c++11 -O0 -g3 -ftrapv -Wstack-protector -fstack-protector-all -Wall -Wextra -Wgnu -Wformat -Wformat-security -Wvla -Wshadow-field -Wswitch -Wthread-safety -Wrange-loop-analysis -Wredundant-decls -Wunused-variable -Wdate-time -Wconditional-uninitialized -Wsign-compare -Woverloaded-virtual -Wunreachable-code-loop-increment -Wno-unused-parameter -Wno-self-assign -Wno-unused-local-typedef -Wno-deprecated-register -Wno-implicit-fallthrough    -fPIE    -Wl,-z,relro -Wl,-z,now -Wl,-z,separate-code -pie     -pthread -lpthread  -o bitcoind bitcoind-bitcoind.o  libbitcoin_server.a libbitcoin_wallet.a libbitcoin_common.a libbitcoin_util.a univalue/libunivalue.la  libbitcoin_consensus.a crypto/libbitcoin_crypto_base.a    leveldb/libleveldb.a crc32c/libcrc32c.a    leveldb/libmemenv.a secp256k1/libsecp256k1.la -L/usr/lib/arm-linux-gnueabihf -lboost_system -lboost_filesystem -lboost_thread -lpthread -ldb_cxx -lminiupnpc -levent_pthreads -levent -levent  -lsqlite3 
libtool: link: clang++ -std=c++11 -O0 -g3 -ftrapv -Wstack-protector -fstack-protector-all -Wall -Wextra -Wgnu -Wformat -Wformat-security -Wvla -Wshadow-field -Wswitch -Wthread-safety -Wrange-loop-analysis -Wredundant-decls -Wunused-variable -Wdate-time -Wconditional-uninitialized -Wsign-compare -Woverloaded-virtual -Wunreachable-code-loop-increment -Wno-unused-parameter -Wno-self-assign -Wno-unused-local-typedef -Wno-deprecated-register -Wno-implicit-fallthrough -fPIE -Wl,-z -Wl,relro -Wl,-z -Wl,now -Wl,-z -Wl,separate-code -pie -pthread -o bitcoind bitcoind-bitcoind.o  -lpthread libbitcoin_server.a libbitcoin_wallet.a libbitcoin_common.a libbitcoin_util.a univalue/.libs/libunivalue.a libbitcoin_consensus.a crypto/libbitcoin_crypto_base.a leveldb/libleveldb.a crc32c/libcrc32c.a leveldb/libmemenv.a secp256k1/.libs/libsecp256k1.a -L/usr/lib/arm-linux-gnueabihf -lboost_system -lboost_filesystem -lboost_thread -lpthread -ldb_cxx -lminiupnpc -levent_pthreads -levent -levent /usr/lib/arm-linux-gnueabihf/libsqlite3.so -pthread
/usr/bin/ld: bitcoind-bitcoind.o: in function `std::chrono::duration<long long, std::ratio<1ll, 1000000ll> > std::chrono::__duration_cast_impl<std::chrono::duration<long long, std::ratio<1ll, 1000000ll> >, std::ratio<1000ll, 1ll>, long long, false, true>::__cast<long long, std::ratio<1ll, 1000ll> >(std::chrono::duration<long long, std::ratio<1ll, 1000ll> > const&)':
/usr/bin/../lib/gcc/arm-linux-gnueabihf/8/../../../../include/c++/8/chrono:168: undefined reference to `__mulodi4'
/usr/bin/ld: libbitcoin_server.a(libbitcoin_server_a-init.o): in function `AppInitParameterInteraction(ArgsManager const&)':
/home/pi/src/bitcoin-core-v0.21/src/init.cpp:1102: undefined reference to `__mulodi4'
/usr/bin/ld: /home/pi/src/bitcoin-core-v0.21/src/init.cpp:1103: undefined reference to `__mulodi4'
/usr/bin/ld: /home/pi/src/bitcoin-core-v0.21/src/init.cpp:1103: undefined reference to `__mulodi4'
/usr/bin/ld: libbitcoin_server.a(libbitcoin_server_a-init.o): in function `AppInitMain(util::Ref const&, NodeContext&, interfaces::BlockAndHeaderTipInfo*)':
/home/pi/src/bitcoin-core-v0.21/src/init.cpp:1532: undefined reference to `__mulodi4'
/usr/bin/ld: libbitcoin_server.a(libbitcoin_server_a-init.o):/home/pi/src/bitcoin-core-v0.21/src/init.cpp:1532: more undefined references to `__mulodi4' follow
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make[2]: *** [Makefile:8343: bitcoind] Error 1
make[2]: Leaving directory '/home/pi/src/bitcoin-core-v0.21/src'
make[1]: *** [Makefile:19347: all-recursive] Error 1
make[1]: Leaving directory '/home/pi/src/bitcoin-core-v0.21/src'
make: *** [Makefile:801: all-recursive] Error 1

Seems similar to this issue, so I'm assuming it has something to do with the choice to use clang - https://github.com/android/ndk/issues/506

The same error while compiling master (eb9a1fe03779bf05062b70f14190cb23ff42b46f) on ODROID-HC1 + Armbian 21.02.3 Focal:
```
  CXXLD    test/fuzz/fuzz
/usr/bin/ld: test/fuzz/fuzz-multiplication_overflow.o: in function `void (anonymous namespace)::TestMultiplicationOverflow<long long>(FuzzedDataProvider&)':
./test/fuzz/multiplication_overflow.cpp:30: undefined reference to `__mulodi4'
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

Clang bug: https://bugs.llvm.org/show_bug.cgi?id=28629

As a workaround, the `--disable-fuzz-binary` option could be passed to the `configure` script.

--- a/build-aux/m4/bitcoin_runtime_lib.m4
+++ b/build-aux/m4/bitcoin_runtime_lib.m4
@@ -1,0 +1,42 @@
+# On some platforms clang builtin implementations
+# require compiler-rt as a runtime library to use.
+#
+# See:
+# - https://bugs.llvm.org/show_bug.cgi?id=28629
+
+m4_define([_CHECK_RUNTIME_testbody], [[
+  bool f(long long x, long long y, long long* p)
+  {
+    return __builtin_mul_overflow(x, y, p);
+  }
+  int main() { return 0; }
+]])
+
+AC_DEFUN([CHECK_RUNTIME_LIB], [
+
+  AC_LANG_PUSH([C++])
+
+  AC_MSG_CHECKING([for __builtin_mul_overflow])
+  AC_LINK_IFELSE(
+    [AC_LANG_SOURCE([_CHECK_RUNTIME_testbody])],
+    [
+      AC_MSG_RESULT([yes])
+      AC_DEFINE([HAVE_BUILTIN_MUL_OVERFLOW], [1], [Define if you have a working __builtin_mul_overflow])
+    ],
+    [
+      ax_check_save_flags="$LDFLAGS"
+      LDFLAGS="$LDFLAGS --rtlib=compiler-rt -lgcc_s"
+      AC_LINK_IFELSE(
+        [AC_LANG_SOURCE([_CHECK_RUNTIME_testbody])],
+        [
+          AC_MSG_RESULT([yes, with additional linker flags])
+          RUNTIME_LDFLAGS="--rtlib=compiler-rt -lgcc_s"
+          AC_DEFINE([HAVE_BUILTIN_MUL_OVERFLOW], [1], [Define if you have a working __builtin_mul_overflow])
+        ],
+        [AC_MSG_RESULT([no])])
+      LDFLAGS="$ax_check_save_flags"
+    ])
+
+  AC_LANG_POP
+  AC_SUBST([RUNTIME_LDFLAGS])
+])

--- a/configure.ac
+++ b/configure.ac
@@ -1773,6 +1773,10 @@ if test x$build_bitcoin_wallet$build_bitcoin_cli$build_bitcoin_tx$build_bitcoin_
   AC_MSG_ERROR([No targets! Please specify at least one of: --with-utils --with-libs --with-daemon --with-gui --enable-bench or --enable-tests])
 fi
 
+if test x$enable_fuzz_binary = xyes; then
+  CHECK_RUNTIME_LIB
+fi
+
 AM_CONDITIONAL([TARGET_DARWIN], [test x$TARGET_OS = xdarwin])
 AM_CONDITIONAL([BUILD_DARWIN], [test x$BUILD_OS = xdarwin])
 AM_CONDITIONAL([TARGET_LINUX], [test x$TARGET_OS = xlinux])

--- a/src/Makefile.test.include
+++ b/src/Makefile.test.include
@@ -200,7 +200,7 @@ if ENABLE_FUZZ_BINARY
 test_fuzz_fuzz_CPPFLAGS = $(AM_CPPFLAGS) $(BITCOIN_INCLUDES)
 test_fuzz_fuzz_CXXFLAGS = $(AM_CXXFLAGS) $(PIE_FLAGS)
 test_fuzz_fuzz_LDADD = $(FUZZ_SUITE_LD_COMMON)
-test_fuzz_fuzz_LDFLAGS = $(FUZZ_SUITE_LDFLAGS_COMMON)
+test_fuzz_fuzz_LDFLAGS = $(FUZZ_SUITE_LDFLAGS_COMMON) $(RUNTIME_LDFLAGS)
 test_fuzz_fuzz_SOURCES = \
  test/fuzz/addition_overflow.cpp \
  test/fuzz/addrdb.cpp \

--- a/src/test/fuzz/multiplication_overflow.cpp
+++ b/src/test/fuzz/multiplication_overflow.cpp
@@ -2,6 +2,10 @@
 // Distributed under the MIT software license, see the accompanying
 // file COPYING or http://www.opensource.org/licenses/mit-license.php.
 
+#if defined(HAVE_CONFIG_H)
+#include <config/bitcoin-config.h>
+#endif
+
 #include <test/fuzz/FuzzedDataProvider.h>
 #include <test/fuzz/fuzz.h>
 #include <test/fuzz/util.h>
@@ -9,14 +13,6 @@
 #include <cstdint>
 #include <string>
 #include <vector>
-
-#if defined(__has_builtin)
-#if __has_builtin(__builtin_mul_overflow)
-#define HAVE_BUILTIN_MUL_OVERFLOW
-#endif
-#elif defined(__GNUC__)
-#define HAVE_BUILTIN_MUL_OVERFLOW
-#endif
 
 namespace {
 template <typename T>

